src/main/kotlin/net/atos/zac/smartdocuments/templates/SmartDocumentsTemplateConverter.kt

```diff
@@ -3,49 +3,147 @@ package net.atos.zac.smartdocuments.templates
+        name = smartDocumentsTemplate.name
+    )
+
+    fun Set<RESTSmartDocumentsTemplateGroup>.toModel(
+        zaakafhandelParameters: ZaakafhandelParameters
+    ): Set<SmartDocumentsTemplateGroup> =
+        this.mapTo(mutableSetOf()) { convertTemplateGroupToModel(it, null, zaakafhandelParameters) }
 
     private fun convertTemplateGroupToModel(
-        group: SmartDocumentsResponseTemplateGroup,
-        parent: SmartDocumentsTemplateGroup?
+        group: RESTSmartDocumentsTemplateGroup,
+        parent: SmartDocumentsTemplateGroup?,
+        zaakafhandelParameterId: ZaakafhandelParameters
     ): SmartDocumentsTemplateGroup {
-        val jpaGroup = createModelTemplateGroup(group, parent)
+        val jpaGroup = createModelTemplateGroup(group, parent, zaakafhandelParameterId)
 
         group.templates?.forEach {
-            jpaGroup.templates.add(createModelTemplate(it))
+            jpaGroup.templates.add(createModelTemplate(it, jpaGroup, zaakafhandelParameterId))
         }
 
-        group.templateGroups?.forEach {
-            jpaGroup.children.add(convertTemplateGroupToModel(it, jpaGroup))
+        group.groups?.forEach {
+            jpaGroup.children.add(convertTemplateGroupToModel(it, jpaGroup, zaakafhandelParameterId))
         }
 
         return jpaGroup
     }
 
     private fun createModelTemplateGroup(
-        smartDocumentsTemplateGroup: SmartDocumentsResponseTemplateGroup,
-        parentGroup: SmartDocumentsTemplateGroup?
+        smartDocumentsTemplateGroup: RESTSmartDocumentsTemplateGroup,
+        parentGroup: SmartDocumentsTemplateGroup?,
+        zaakafhandelParams: ZaakafhandelParameters
     ) = SmartDocumentsTemplateGroup().apply {
         smartDocumentsId = smartDocumentsTemplateGroup.id
+        zaakafhandelParameters = zaakafhandelParams
         name = smartDocumentsTemplateGroup.name
         parent = parentGroup
         creationDate = ZonedDateTime.now()
     }
 
     private fun createModelTemplate(
-        smartDocumentsTemplate: SmartDocumentsResponseTemplate
+        smartDocumentsTemplate: RESTSmartDocumentsTemplate,
+        parentGroup: SmartDocumentsTemplateGroup,
+        zaakafhandelParams: ZaakafhandelParameters
     ) = SmartDocumentsTemplate().apply {
         smartDocumentsId = smartDocumentsTemplate.id
+        zaakafhandelParameters = zaakafhandelParams
         name = smartDocumentsTemplate.name
+        templateGroup = parentGroup
         creationDate = ZonedDateTime.now()
     }
+
+    fun Set<SmartDocumentsTemplateGroup>.toREST(): Set<RESTSmartDocumentsTemplateGroup> =
+        this.mapTo(mutableSetOf()) { convertTemplateGroupToREST(it) }
+
+    private fun convertTemplateGroupToREST(
+        group: SmartDocumentsTemplateGroup
+    ): RESTSmartDocumentsTemplateGroup {
+        val restTemplateGroup = createRESTTemplateGroup(group)
+
+        if (group.templates.isNotEmpty()) {
+            restTemplateGroup.templates = mutableSetOf<RESTSmartDocumentsTemplate>().apply {
```

Comment:
This conversion should also be written in a more functional way, then it'll be a one liner. Like the other suggestions I made, and similarly with the below code.